# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.
`! git clone https://www.github.com/DS3001/assignment2`

In [1]:
! git clone https://www.github.com/DS3001/assignment2

Cloning into 'assignment2'...
remote: Enumerating objects: 36, done.
remote: Counting objects: 100% (36/36), done.
remote: Compressing objects: 100% (30/30), done.
remote: Total 36 (delta 9), reused 29 (delta 6), pack-reused 0
Receiving objects: 100% (36/36), 5.47 MiB | 28.71 MiB/s, done.
Resolving deltas: 100% (9/9), done.


In [2]:
import numpy as np
import pandas as pd

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [2]:
! pwd

/content


1.

In [3]:
# first I loaded the data
airbnb = pd.read_csv('./assignment2/data/airbnb_hw.csv')

In [4]:
# then I examined the price variable -- I saw that it was getting treated as a non-numeric variable
print(airbnb['Price'].describe())
# I looked at the unique values in 'Price' to see if I could tell what was causing that
print(airbnb['Price'].unique())

count     30478
unique      511
top         150
freq       1481
Name: Price, dtype: object
['145' '37' '28' '199' '549' '149' '250' '90' '270' '290' '170' '59' '49'
 '68' '285' '75' '100' '150' '700' '125' '175' '40' '89' '95' '99' '499'
 '120' '79' '110' '180' '143' '230' '350' '135' '85' '60' '70' '55' '44'
 '200' '165' '115' '74' '84' '129' '50' '185' '80' '190' '140' '45' '65'
 '225' '600' '109' '1,990' '73' '240' '72' '105' '155' '160' '42' '132'
 '117' '295' '280' '159' '107' '69' '239' '220' '399' '130' '375' '585'
 '275' '139' '260' '35' '133' '300' '289' '179' '98' '195' '29' '27' '39'
 '249' '192' '142' '169' '1,000' '131' '138' '113' '122' '329' '101' '475'
 '238' '272' '308' '126' '235' '315' '248' '128' '56' '207' '450' '215'
 '210' '385' '445' '136' '247' '118' '77' '76' '92' '198' '205' '299'
 '222' '245' '104' '153' '349' '114' '320' '292' '226' '420' '500' '325'
 '307' '78' '265' '108' '123' '189' '32' '58' '86' '219' '800' '335' '63'
 '229' '425' '67' '87' '1,200' '15

In [6]:
# I believe that the commas in values over 1,000 are making this variable get treated as a string

# So, I removed the commas
airbnb['Price'] = airbnb['Price'].str.replace(',', '')
# and converted the variable to numeric
airbnb['Price'] = pd.to_numeric(airbnb['Price'])

In [7]:
# Then I checked to make sure that the variable was now numeric
print(airbnb['Price'].describe(), '\n')

# and I made sure that there wee no null or NA values in the variable
print(airbnb['Price'].isnull().sum() + airbnb['Price'].isna().sum())

# since there are no null values and since the variable is now numeric, I believe it is fully cleaned!

count    30478.000000
mean       163.589737
std        197.785454
min         10.000000
25%         80.000000
50%        125.000000
75%        195.000000
max      10000.000000
Name: Price, dtype: float64 

0


2.

In [8]:
# first I loaded the data
sharks = pd.read_csv('./assignment2/data/sharks.csv')

<ipython-input-8-ca3a298e4d52>:2: DtypeWarning: Columns (10,17,18,19,20,21,24) have mixed types. Specify dtype option on import or set low_memory=False.
  sharks = pd.read_csv('./assignment2/data/sharks.csv')


In [9]:
# then I examined the type of the variable
print(sharks['Type'].describe(), '\n')
# and the different values it has
sharks['Type'].unique()

count           6457
unique            12
top       Unprovoked
freq            4716
Name: Type, dtype: object 



array(['Unprovoked', 'Provoked', 'Questionable', 'Watercraft',
       'Unconfirmed', 'Unverified', 'Invalid', 'Under investigation',
       'Boating', 'Sea Disaster', nan, 'Boat', 'Boatomg'], dtype=object)

In [10]:
# Several of the values seem similar, so I'm going to combine groups of similar values into one
# First, "Boat" will represent the "Boating", "Boat", and "Boatomg" values (I believe that Boatomg is a typo)
# "Unverified" will represent the "Unverified", "Questionable", and "Unconfirmed" values
# I kept "under investifation" as its own category, because that indicates that the status of the attack might be updated later,
# whereas other "unconfirmed" cases might not get investigated
# a watercraft could be different enough from boat that I kept it as its own category (a jetki is a watercraft, for ex)

def consolodate_attack_type_values(original_value):
  if original_value == "Unconfirmed" or original_value == "Questionable":
    return "Unverified"
  if original_value == "Boating" or original_value == "Boatomg":
    return "Boat"
  if original_value == "Under investigation":
    return "Under Investigation" # I wanted to make the value name capital to be consistent
  return original_value

sharks['Type'] = sharks['Type'].apply(consolodate_attack_type_values)

In [11]:
# check that the transformation worked
sharks['Type'].unique()

array(['Unprovoked', 'Provoked', 'Unverified', 'Watercraft', 'Invalid',
       'Under Investigation', 'Boat', 'Sea Disaster', nan], dtype=object)

In [12]:
# see how many observations do not have an attack type
sharks['Type'].isna().sum()

5

In [13]:
# there are only five observations with a nan value, so it would make sense to simply remove them from the dataset
# however, I want to practice other ways of dealing with nan values
# since I don't know why the Type variable is missing for these observations, I will create a dummy variable to show which observations are missing it

sharks['Type_NAN'] = sharks['Type'].isna()

In [18]:
sharks["Type_NAN"].describe()

count      6462
unique        2
top       False
freq       6457
Name: Type_NAN, dtype: object

In [15]:
# Now I have dealt with missing values and have consolidated like values, so I think this variable is ready to use

3.

In [67]:
# load data -- this code modified from lecture notes
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
pretrial = pd.read_csv(url,low_memory=False) # Pandas downloads and loads the .csv file for you

In [68]:
# inspect the relevant variable
print(pretrial['WhetherDefendantWasReleasedPretrial'].describe(), '\n')
print(pretrial['WhetherDefendantWasReleasedPretrial'].unique(), '\n')
print(pretrial['WhetherDefendantWasReleasedPretrial'].value_counts())

count    22986.000000
mean         0.845428
std          0.477276
min          0.000000
25%          1.000000
50%          1.000000
75%          1.000000
max          9.000000
Name: WhetherDefendantWasReleasedPretrial, dtype: float64 

[9 0 1] 

1    19154
0     3801
9       31
Name: WhetherDefendantWasReleasedPretrial, dtype: int64


In [69]:
# first I'm going to rename this variable, because the current name is so long that it's unwieldy
missing_values = pretrial['WhetherDefendantWasReleasedPretrial'] == 9
pretrial = pretrial.rename(columns = {'WhetherDefendantWasReleasedPretrial':'Released'})

In [70]:
# and I'll check that the rename worked:
# first, check that the old column name no longer exists
print('WhetherDefendantWasReleasedPretrial' in pretrial.columns)
# then check that the variable appears under its new name
print(pretrial['Released'].describe())

False
count    22986.000000
mean         0.845428
std          0.477276
min          0.000000
25%          1.000000
50%          1.000000
75%          1.000000
max          9.000000
Name: Released, dtype: float64


In [71]:
# because 9 is the missing value code, I will replace all the 9s in the dataset with np.nan
missing_values = pretrial['Released'] == 9
pretrial.loc[missing_values, 'Released'] = np.nan
# I'll then confirm that the replacement worked
print(pretrial['Released'].value_counts())

1.0    19154
0.0     3801
Name: Released, dtype: int64


In [72]:
# finally, I'll make a dummy value to encode which rows have missing data for the 'WhetherDefendantWasReleasedPretrial' variable
pretrial['Released_NAN'] = pretrial['Released'].isna()
print(pretrial['Released_NAN'].value_counts())

False    22955
True        31
Name: Released_NAN, dtype: int64


4.
Missing values, not at random: For the pretrial data covered in the lecture, clean the ImposedSentenceAllChargeInContactEvent variable as well as you can, and explain the choices you make. (Hint: Look at the SentenceTypeAllChargesAtConvictionInContactEvent variable.)

In [94]:
pretrial2 = pretrial.copy(deep=True)

In [95]:
# first, I examined the imposed sentence variable and saw that it was of type "object"
print(pretrial2['ImposedSentenceAllChargeInContactEvent'].describe(), '\n')
print(pretrial2['ImposedSentenceAllChargeInContactEvent'].unique())

count     22986
unique      484
top            
freq       9053
Name: ImposedSentenceAllChargeInContactEvent, dtype: object 

[' ' '60' '12' '.985626283367556' '36' '6' '24' '5.91375770020534' '120'
 '72' '11.9917864476386' '0' '2.95687885010267' '84' '108' '300' '240'
 '180' '4' '96' '2' '54' '.328542094455852' '44' '5' '115' '132' '48'
 '258' '34' '76' '.164271047227926' '.131416837782341' '111' '9' '3'
 '1.97125256673511' '36.9856262833676' '.0657084188911704'
 '35.4928131416838' '106.492813141684' '8' '35' '18.3141683778234' '480'
 '32' '93' '234' '732' '1.16427104722793' '4.6570841889117' '21' '7'
 '4.49281314168378' '18' '600' '43.1642710472279' '179' '52' '30' '20'
 '192' '702' '14' '55' '53' '11.9055441478439' '114' '35.0061601642711'
 '68' '.657084188911704' '46.6242299794661' '102' '65' '200' '57'
 '24.3285420944559' '12.1642710472279' '117' '81.4928131416838'
 '22.4928131416838' '1980' '3.6570841889117' '56' '10' '2.79260780287474'
 '1' '47' '22' '1500' '40' '284' '11' '118'

In [96]:
# one of the values for ImposedSentence was a space, which would not be able to be converted to a number
# I checked to see how many times this value occurred in the variable
print((pretrial2['ImposedSentenceAllChargeInContactEvent'] == ' ').sum())

9053


In [97]:
# next, I converted the variable to a numeric type and examined it
pretrial2['ImposedSentenceAllChargeInContactEvent'] = pd.to_numeric(pretrial2['ImposedSentenceAllChargeInContactEvent'], errors='coerce')
print(pretrial2['ImposedSentenceAllChargeInContactEvent'].describe(), '\n')
print(pretrial2['ImposedSentenceAllChargeInContactEvent'].value_counts())

count    13933.000000
mean        19.899671
std         63.571627
min          0.000000
25%          0.000000
50%          2.000000
75%         12.000000
max       2208.000000
Name: ImposedSentenceAllChargeInContactEvent, dtype: float64 

0.000000     4953
12.000000    1404
0.985626     1051
6.000000      809
3.000000      787
             ... 
49.971253       1
57.034908       1
79.926078       1
42.164271       1
1.657084        1
Name: ImposedSentenceAllChargeInContactEvent, Length: 483, dtype: int64


In [98]:
# I checked to make sure that the spaces (and only the spaces) became nans
pretrial2["ImposedSentenceAllChargeInContactEvent"].isna().sum()

9053

In [99]:
# I can see that there are a lot of sentences of zero months, and there are also 9053 observations with no sentence
# I'll create a dummy variable for when there is a sentence of zero months
pretrial2["Sentence_Zero"] = pretrial2["ImposedSentenceAllChargeInContactEvent"] == 0
# and one for when there is no sentence
pretrial2["Sentence_NA"] = pretrial2["ImposedSentenceAllChargeInContactEvent"].isna()

In [100]:
# then, I'll cross-tabulate these dummy variablea with the SentenceType variable
print(pd.crosstab(pretrial2['Sentence_Zero'], pretrial2['SentenceTypeAllChargesAtConvictionInContactEvent']), '\n')
print(pd.crosstab(pretrial2['Sentence_NA'], pretrial2['SentenceTypeAllChargesAtConvictionInContactEvent']))

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
Sentence_Zero                                                               
False                                             3767  4299  914  8779  274
True                                              4953     0    0     0    0 

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
Sentence_NA                                                                 
False                                             8720  4299  914     0    0
True                                                 0     0    0  8779  274


In [101]:
# If the sentence type is 9, the sentence should be np.nan, because in that case, the data about sentence length is unknown
# Similarly, if the charges against the person were dropped, then of course there is no sentence to speak about
# So, it makes sense that all the sentences for people with sentence types 4 and 9 are NaN

In [102]:
# looking at the codebook, sentence type 0 means that the person was put on probation/not incarcerated
# this indicates a more minor crime, so it makes sense that this is the only category
# where people received a sentence of zero months
# in this case, a sentence was given, and we have information about it, but it was chosen to be zero months
# so, I think this value should NOT be replaced with NaN

# I examined the sentences given to people with this sentence type
pretrial2.loc[ pretrial2['SentenceTypeAllChargesAtConvictionInContactEvent'] == 0, 'ImposedSentenceAllChargeInContactEvent'].value_counts()

0.000000      4953
0.985626       765
12.000000      586
3.000000       442
6.000000       372
              ... 
158.000000       1
23.663244        1
227.000000       1
59.663244        1
0.459959         1
Name: ImposedSentenceAllChargeInContactEvent, Length: 176, dtype: int64

In [103]:
# the one thing I noticed was that some people still got a sentence despite not being incarcerated
# I think it would make more sense to set these people's sentence length as 0 (indicating that they were not sentenced to jail time)
# and then represent their probation time in its own variable, ProbationTime

# I made a new column for this variable, copying over the sentence times
pretrial2['ProbationTime'] = pretrial2['ImposedSentenceAllChargeInContactEvent'].copy()
# then made the probation time NaN for any sentence type that was not 0 (did not involve probation)
pretrial2.loc[ pretrial2['SentenceTypeAllChargesAtConvictionInContactEvent'] != 0, 'ProbationTime'] = np.nan
# finally, I reflected that all the people with sentence type 0 did not receive any jail time
pretrial2.loc[ pretrial2['SentenceTypeAllChargesAtConvictionInContactEvent'] == 0, 'ImposedSentenceAllChargeInContactEvent'] = 0.0

In [104]:
# finally, I made sure that the changes had been applied by re-running/running more cross tabulations
pretrial2["Sentence_Zero"] = pretrial2["ImposedSentenceAllChargeInContactEvent"] == 0
pretrial2["Sentence_NA"] = pretrial2["ImposedSentenceAllChargeInContactEvent"].isna()
pretrial2["Probation_Zero"] = pretrial2["ProbationTime"] == 0
pretrial2["Probation_NA"] = pretrial2["ProbationTime"].isna()

print(pd.crosstab(pretrial2['Sentence_Zero'], pretrial2['SentenceTypeAllChargesAtConvictionInContactEvent']), '\n')
print(pd.crosstab(pretrial2['Sentence_NA'], pretrial2['SentenceTypeAllChargesAtConvictionInContactEvent']), '\n')
print(pd.crosstab(pretrial2['Probation_Zero'], pretrial2['SentenceTypeAllChargesAtConvictionInContactEvent']), '\n')
print(pd.crosstab(pretrial2['Probation_NA'], pretrial2['SentenceTypeAllChargesAtConvictionInContactEvent']))

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
Sentence_Zero                                                               
False                                                0  4299  914  8779  274
True                                              8720     0    0     0    0 

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
Sentence_NA                                                                 
False                                             8720  4299  914     0    0
True                                                 0     0    0  8779  274 

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
Probation_Zero                                                              
False                                             3767  4299  914  8779  274
True                                              4953     0    0     0    0 

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     

In [105]:
# looks good!

**Q3.** This question provides some practice doing exploratory data analysis and visualization.

The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `./data/college_completion.csv` data with Pandas.
2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see.
4. For `grad_100_value`, create a histogram, kernel density plot, boxplot, and statistical description.
5. For `grad_100_value`, create a grouped kernel density plot by `control` and by `level`. Describe what you see. Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `grad_100_value` by `level` and `control`. Which institutions appear to have the best graduation rates?
6. Create a new variable, `df['levelXcontrol']=df['level']+', '+df['control']` that interacts level and control. Make a grouped kernel density plot. Which institutions appear to have the best graduation rates?
7. Make a kernel density plot of `aid_value`. Notice that your graph is "bi-modal", having two little peaks that represent locally most common values. Now group your graph by `level` and `control`. What explains the bi-modal nature of the graph? Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `aid_value` by `level` and `control`.
8. Make a scatterplot of `grad_100_value` by `aid_value`. Describe what you see. Now make the same plot, grouping by `level` and then `control`. Describe what you see. For which kinds of institutions does aid seem to increase graduation rates?

**Q4.** This question uses the Airbnb data to practice making visualizations.

  1. Load the `./data/airbnb_hw.csv` data with Pandas. You should have cleaned the `Price` variable in question 2, and you'll need it later for this question.
  2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
  3. Cross tabulate `Room Type` and `Property Type`. What patterns do you see in what kinds of rentals are available? For which kinds of properties are private rooms more common than renting the entire property?
  4. For `Price`, make a histogram, kernel density, box plot, and a statistical description of the variable. Are the data badly scaled? Are there many outliers? Use `log` to transform price into a new variable, `price_log`, and take these steps again.
  5. Make a scatterplot of `price_log` and `Beds`. Describe what you see. Use `.groupby()` to compute a desciption of `Price` conditional on/grouped by the number of beds. Describe any patterns you see in the average price and standard deviation in prices.
  6. Make a scatterplot of `price_log` and `Beds`, but color the graph by `Room Type` and `Property Type`. What patterns do you see? Compute a description of `Price` conditional on `Room Type` and `Property Type`. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?
  7. We've looked a bit at this `price_log` and `Beds` scatterplot. Use seaborn to make a `jointplot` with `kind=hex`. Where are the data actually distributed? How does it affect the way you think about the plots in 5 and 6?

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

**Q6.** Open the `./data/CBO_data.pdf` file. This contains tax data for 2019, explaining where the money comes from that the U.S. Federal Government Spends in terms of taxation on individuals/families and payroll taxes (the amount that your employer pays in taxes on your wages).

For some context, the Federal government ultimately spent about $4.4 trillion in 2019, which was 21% of GDP (the total monetary value of all goods and services produced within the United States). Individual Income Taxes is the amount individuals pay on their wages to the Federal government, Corporate Income Taxes is the taxes individuals pay on capital gains from investment when they sell stock or other financial instruments, Payroll Taxes is the tax your employer pays on your wages, Excises and Customs Duties are taxes on goods or services like sin taxes on cigarettes or alcohol, and Estate and Gift Taxes are taxes paid on transfers of wealth to other people.

1. Get the Millions of Families and Billions of Dollars data into a .csv file and load it with Pandas.
2. Create a bar plot of individual income taxes by income decile. Explain what the graph shows. Why are some values negative?
3. Create a bar plot of Total Federal Taxes by income decile. Which deciles are paying net positive amounts, and which are paying net negative amounts?
4. Create a stacked bar plot for which Total Federal Taxes is grouped by Individual Income Taxes, Payroll Taxes, Excises and Customs Duties, and Estate and Gift Taxes. How does the share of taxes paid vary across the adjusted income deciles? (Hint: Are these the kind of data you want to melt?)
5. Below the Total line for Millions of Families and Billions of Dollars, there are data for the richest of the richest families. Plot this alongside the bars for the deciles above the Total line. Describe your results.
6. Get the Percent Distribution data into a .csv file and load it with Pandas. Create a bar graph of Total Federal Taxes by income decile.
7. A tax system is progressive if higher-income and wealthier individuals pay more than lower-income and less wealthy individuals, and it is regressive if the opposite is true. Is the U.S. tax system progressive in terms of amount paid? In terms of the percentage of the overall total?
8. Do the rich pay enough in taxes? Defend your answer.